### DISCUSSION TITLE
Wasm in Web Worker, ES Modules

### DISCUSSION THREAD

**Discussion by tysen** [wasm-bindgen/wasm-bindgen#3519] [Q&A] [answered]
### Summary

How can I use ES Modules in a Web Worker to import wasm-loading JS?

### Additional Details

The [example](https://rustwasm.github.io/wasm-bindgen/examples/wasm-in-web-worker.html) assumes you want to use `--target no-modules` because of a lack of support for modules in Firefox. However, Firefox 114 [added support for ES Modules in workers](https://www.mozilla.org/en-US/firefox/114.0/releasenotes/).

Additionally, while this example's title is "Wasm in Web Worker", it would be more aptly titled "Web Worker in Wasm" as the worker itself is spawned in the Rust that's compiled to wasm and `index.js` just executes the wasm entrypoint function (which itself executes `worker.js`). This seems convoluted to me. Can't I do something like this?

`index.js`
```javascript
let w = new Worker("worker.js", { "type": "module" });
w.postMessage(..);
```

`worker.js`
```javascript
import { add } from "wasm.js"
self.onmessage = async e => {
  add(e.a, e.b);
};
```

`wasm.js`
`auto generated using wasm-pack build --target web`
```javascript
export function add(a, b) {
  const ret = wasm.add(a, b);
  return ret >>> 0;
}

function initSync(module) {
  // contents of autogen func omitted for brevity
}

async function __wbg_init(input) {
  // contents of autogen func omitted for brevity
}

export { initSync }
export default __wbg_init;
```

`lib.rs`
```rust
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub fn add(a: u32, b: u32) -> u32 {
    a + b
}
```

Certainly this is wrong. I know that there is logic missing in `worker.js` - I don't know how/whether to use either `initSync` or `__wbg_init`. But I think this setup illustrates my questions:

1. How can I use ES Modules in a Web Worker to import wasm-loading JS?
2. Is the convoluted nature of the non-module method in the example inescapable? I would like to "instantiate a Worker that uses wasm" not "use wasm to instantiate a Worker".

**Comment by daxpedda** [comment, marked as answer]
Answered on [Discord](https://discord.com/channels/442252698964721669/1129130697441943613/1129408066383794338):
> You can use either.
After spawning the worker, you have to send over the module and use that to initialize the instance.
https://docs.rs/wasm-bindgen/0.2.87/src/wasm_bindgen/lib.rs.html#1355-1364
https://docs.rs/wasm-bindgen/0.2.87/wasm_bindgen/fn.memory.html
>
> By calling `initSync(module, memory)`
